### ISSUE TITLE
get_params returns different values after saving and loading the model

### ISSUE THREAD

**Issue by BenjaminBossan** [dmlc/xgboost#8596]
I'm using the sklearn API for xgboost and noticed that for some model parameters I tested, the values would change after saving and loading the models. However, this was only true when accessing the parameters via `get_params`, not via normal attribute access.

I haven't tested each existing parameter, the ones I found were `tree_method` and `gpu_id`, both with default values `None`. For `tree_method`, the value changes from `'exact'` to `'auto'`, and for `gpu_id` from 0 to -1. Whether this has any practical implications, I don't know.

Below is some code to reproduce the issue.

```
machine: Linux-5.15.0-56-generic-x86_64-with-glibc2.35
python: 3.10.8
xgboost: 1.7.2
```

```python
>>> # value of tree_method changes
>>> import xgboost
>>> estimator = xgboost.XGBClassifier(tree_method=None)
>>> X, y = [[0, 1], [2, 3]], [0, 1]
>>> estimator.fit(X, y)
XGBClassifier(...)
>>> print(estimator.tree_method)
None
>>> print(estimator.get_params()["tree_method"])
exact
>>> # after save/load roundtrip, tree_method's get_params value changes from exact to auto
>>> import tempfile
>>> tmp_file = f"{tempfile.mkdtemp()}.ubj"
>>> estimator.save_model(tmp_file)
>>> estimator.load_model(tmp_file)
>>> print(estimator.tree_method)
None
>>> print(estimator.get_params()["tree_method"])
auto  # <== was 'exact'

>>> # a similar thing happens with gpu_id
>>> estimator = xgboost.XGBClassifier(tree_method='gpu_hist', booster='gbtree')
>>> estimator.fit(X, y)
XGBClassifier(...)
>>> print(estimator.gpu_id)
None
>>> print(estimator.get_params()["gpu_id"])
0
>>> # after save/load roundtrip, gpu_id's get_params value changes from 0 to -1
>>> estimator.save_model(tmp_file)
>>> # for gpu_id, the estimator needs to be re-initialized for the effect to occur
>>> estimator = xgboost.XGBClassifier()
>>> estimator.load_model(tmp_file)
>>> print(estimator.gpu_id)
None
>>> print(estimator.get_params()["gpu_id"])
-1  # <== was 0
```

**Comment by trivialfis** [comment]
Hi, this is expected for some complicated reason.

Longer explanation:
XGBoost distinguishes "model parameter" and "hyper-parameter" used for training. `tree_method` and `gpu_id` are hyper-parameters that are used to generate the model. For a neural network model, the nn is the model and the learning rate is a hyper-parameter. See https://xgboost.readthedocs.io/en/stable/tutorials/saving_model.html for details. `save_model` for xgboost is supposed to save only the model, not the hyper-parameter that used to generate the model. Parameters like `n_jobs` and `gpu_id` are more obvious and are not supposed to be saved in a model since the model should be used with any hardware without restriction.

However, for the sklearn estimator interface, parameters don't have such differences, and so there are some inconsistencies for how to preserve the parameters. We save the python attributes with the current version of XGBoost (which we intend to remove in the future and save only the model), as an odd result, after loading the model from a round-trip, the python parameter you passed in originally takes effect again.

In your example, the `tree_method` is specified as `None` in `estimator = xgboost.XGBClassifier(tree_method=None)`, which means `auto` for XGBoost. With a small input, XGBoost chooses `exact` to be its tree method and returns the string `exact` after auto configuration. After loading back the model, the python attribute `tree_method=None` is restored, and so `get_params()["tree_method"]` returns what's passed in: `None`, instead of the configured `exact`.

Similar for `gpu_id`, `None` is default, which implies `-1` for CPU. `0` is configured for using the first GPU for `gpu_hist`. After saving and loading the model, the default `None` from python attribute is restored.

**Comment by adrinjalali** [comment]
I see. Thanks for the explanation. Although, this doesn't make the sklearn estimators strictly scikit-learn compatible.

**Comment by trivialfis** [comment]
Let me take another look into this and see if there's way to improve it.

**Comment by adrinjalali** [comment]
@trivialfis feel free to ping me if you think I could be of any help there :)

**Comment by BenjaminBossan** [comment]
> In your example, the `tree_method` is specified as `None` in `estimator = xgboost.XGBClassifier(tree_method=None)`, which means `auto` for XGBoost. With a small input, XGBoost chooses `exact` to be its tree method and returns the string `exact` after auto configuration. After loading back the model, the python attribute `tree_method=None` is restored, and so `get_params()["tree_method"]` returns what's passed in originally: `None`, instead of the configured `exact`.

I think how sklearn would solve this is that the `__init__` parameters like `tree_method` and `gpu_id` are treated as immutable. If, during the `fit` call, it is determined that, e.g., `'exact'` is the tree method of choice, a new attribute `self.tree_method_ = 'exact'` is stored (with trailing underscore). Later in code, only `self.tree_method_` is used.

**Comment by trivialfis** [comment]
I opened a PR for hiding the internally configured value for sklearn estimators: https://github.com/dmlc/xgboost/pull/8634 .
